src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2749,13 +2813,564 @@ def query(self, filters=None, options=None):
+            if not dataset['locked']:
+                verrors.add('id', f'{id} dataset is not locked')
+            elif not dataset['encryption_root'] != id:
+                verrors.add('id', 'Only encryption roots can be unlocked')
+            else:
+                if not bool(self.encrypted_roots_query_db([['name', '=', id]])) and id not in keys_supplied:
+                    verrors.add('unlock_options.datasets', f'Please specify key for {id}')
+
+        verrors.check()
+
+        locked_datasets = []
+        datasets = self.query_encrypted_datasets(
+            id.split('/', 1)[0], {
+                'recursive': True, 'load_db': True, 'full_output': True, 'key_loaded': False
+            }
+        )
+        for name, ds in datasets.items():
+            datasets[name] = {'key': keys_supplied.get(name) or ds['encryption_key'], **ds}
+            if ds['locked'] and id != name and id.startswith(name):
+                # This ensures that `id` has locked parents and they should be unlocked first
+                locked_datasets.append(name)
+
+        if locked_datasets:
+            raise CallError(f'{id} has locked parents {",".join(locked_datasets)} which must be unlocked first')
+
+        failed = defaultdict(lambda: dict({'error': None, 'skipped': []}))
+        unlocked = []
+        for name in sorted(
+            filter(
+                lambda n: n and n.startswith(id) and datasets[n]['locked'],
+                (datasets if options['recursive'] else [datasets.get(id)])
+            ),
+            key=lambda v: v.count('/')
+        ):
+            if not datasets[name]['key']:
+                failed[name]['error'] = 'Missing key'
+                continue
+
+            skip = False
+            for i in range(1, name.count('/')):
+                check = name.rsplit('/', i)[0]
+                if check in failed:
+                    failed[check]['skipped'].append(name)
+                    skip = True
+                    break
+
+            if skip:
+                continue
+
+            try:
+                self.middleware.call_sync(
+                    'zfs.dataset.load_key', name, {'key': datasets[name]['key'], 'mount': True, 'recursive': True}
+                )
+            except CallError as e:
+                failed[name]['error'] = 'Invalid Key' if 'incorrect key provided' in str(e).lower() else str(e)
+            else:
+                unlocked.append(name)
+
+        if options['toggle_attachments'] and not failed:
+            j = self.middleware.call_sync('pool.dataset.toggle_attachments', self.__attachments_path(dataset), True)
+            j.wait_sync()
+
+        return {'unlocked': unlocked, 'failed': failed}
+
+    @private
+    @job(lock=lambda args: f'toggle_attachments_{args[0]}')
+    async def toggle_attachments(self, path, enabled=True):
+        for delegate in self.attachment_delegates:
+            delegate.toggle((await delegate.query(path, enabled)), enabled)
```

Comment:
I might been found guilty in not documenting `FSAttachmentDelegate` so I did it now.

Your code disables attachment that are already disabled or enables ones that are already enabled. That has no sense for me. What did you want to do?